util/patch/fake/client.go

```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fake
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+// NewFakeClientWithScheme creates a new fake client with the given scheme
+// for testing.
+// You can choose to initialize it with a slice of runtime.Object; all the objects with be given
+// a fake ResourceVersion="1" so it will be possible to use optimistic lock when patching conditions.
+func NewFakeClientWithScheme(clientScheme *runtime.Scheme, initObjs ...runtime.Object) client.Client {
+	for _, obj := range initObjs {
+		accessor, err := meta.Accessor(obj)
+		if err != nil {
+			panic(fmt.Errorf("failed to get accessor for object: %v", err))
+		}
+
+		if accessor.GetResourceVersion() == "" {
+			accessor.SetResourceVersion("1")
+		}
```

Comment:
I'm wondering if it might be a good idea to also create a deep copy of any passed in resources as well, which could help avoid potential issues where someone forgets to make a deep copy of a resource that is used in multiple tests across mutliple fake clients?

Reply:
Uhhhmm
when I call client.Create I do expect side effects, and NewFakeClientWithScheme is just an helper for doing NewClient and client.Create....

Reply:
It gets complicated, though...  If you call `client.Create` you expect the side effect to be that the resource reflects the state immediately after the Create call.

With the Fake client, and passing in the resources to `NewFakeClientWithScheme` I believe you may also get changes made from other operations against the backing resource if you don't provide it a deep copy. So if you designed a test where you:
- create a resource to use for initializing and comparison later, let's call it `preResource`
- pass the pointer to `preResource` resource in directly to `NewFakeClientWiithScheme`
- call a function/method that is meant to modify the resource represented by `preResource`, let's call it `storedResource`
- get a copy of `storedResource` using `client.Get, let's call this copy `postResource`

Because `storedResource` and `preResource` are pointing to the same underlying resource, someone building a test could mistakenly attempt to make comparisons using `preResource` against `postResource` under the assumption that `preResource` actually represents the state of the resource prior to calling the function that modifies `storedResource`, however it actually represents the state of `storedResource`.

Not sure if that helps with clarifying my concern.

Reply:
Yeah, @detiber is correct here — we need to make sure to not modify the caller object during init. I hope in the future we only use Create and avoid using any init objects.